prediction_market_agent/agents/ofvchallenger_agent/deploy.py

```diff
@@ -1,21 +1,146 @@
+from datetime import timedelta
+
 from prediction_market_agent_tooling.deploy.agent import DeployableAgent
+from prediction_market_agent_tooling.gtypes import ChecksumAddress, xdai_type
+from prediction_market_agent_tooling.loggers import logger
 from prediction_market_agent_tooling.markets.markets import MarketType
+from prediction_market_agent_tooling.markets.omen.data_models import (
+    OmenMarket,
+    RealityResponse,
+)
+from prediction_market_agent_tooling.markets.omen.omen_resolving import (
+    Resolution,
+    omen_submit_answer_market_tx,
+)
+from prediction_market_agent_tooling.markets.omen.omen_subgraph_handler import (
+    OmenSubgraphHandler,
+)
 from prediction_market_agent_tooling.tools.langfuse_ import observe
+from prediction_market_agent_tooling.tools.utils import utcnow
+from pydantic import BaseModel
+from web3 import Web3
 
+from prediction_market_agent.agents.ofvchallenger_agent.ofv_resolver import (
+    ofv_answer_binary_question,
+)
+from prediction_market_agent.agents.replicate_to_omen_agent.omen_resolve_replicated import (
+    claim_all_bonds_on_reality,
+)
 from prediction_market_agent.utils import APIKeys
 
 OFV_CHALLENGER_TAG = "ofv_challenger"
+OFV_CHALLENGER_EOA_ADDRESS = Web3.to_checksum_address(
+    "0x03aEEE267A985ab9835c025377F0D8209901a928"
+)
+OFV_CHALLENGER_SAFE_ADDRESS = Web3.to_checksum_address(
+    "0x9D0260500ba7b068b5b0f4AfA9F8864eBc0B059a"
+)
+CHALLENGE_BOND = xdai_type(10)
+# We benchmarked OFV only against Olas market-creators.
+MARKET_CREATORS_TO_CHALLENGE: list[ChecksumAddress] = [
+    # Olas market-creator 0.
+    Web3.to_checksum_address("0x89c5cc945dd550bcffb72fe42bff002429f46fec"),
+    # Olas market-creator 1.
+    Web3.to_checksum_address("0xffc8029154ecd55abed15bd428ba596e7d23f557"),
+]
+
+
+class Challenge(BaseModel):
+    old_responses: list[RealityResponse]
+    new_resolution: Resolution
 
 
 class OFVChallengerAgent(DeployableAgent):
     def run(self, market_type: MarketType) -> None:
         if market_type != MarketType.OMEN:
             raise RuntimeError("Can challenge only Omen.")
 
-        self.challenge()
+        api_keys = APIKeys()
+        self.challenge(api_keys)
 
     @observe()
-    def challenge(self) -> None:
+    def challenge(self, api_keys: APIKeys) -> None:
         self.langfuse_update_current_trace(tags=[OFV_CHALLENGER_TAG])
-        APIKeys()
-        # TODO: Coming in the next PR.
+
+        # Claim the bonds as first thing, to have funds for the new challenges.
+        claim_all_bonds_on_reality(api_keys)
+
+        markets_open_for_answers = OmenSubgraphHandler().get_omen_binary_markets(
```

Comment:
does this include markets that are resolved as invalid? if so do we want to also challenge these?

Reply:
> does this include markets that are resolved as invalid?

Yes it should, by default we don't have any filter for invalid answers in place, only if we ask for resolved markets:

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/1824de9a-9af1-4595-acba-905f4c45ee6b)

>  if so do we want to also challenge these?

I think we do. If Olas market-creator resolve it as Invalid, it doesn't mean it's true (same as with Yes/No resolutions).

However, currently if OFV can not find the answer, we won't challenge it:

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/6ecb4436-abfb-404f-9934-8be6f77ea77b)

Should we challenge to Invalid if `factuality` is `None`?

Reply:
Yes, I think that makes sense. When you ran the benchmark to compare current resolutions to OFV's, did this include any invalid markets? Wondering if it's a bad idea to do this if we don't know how good it is as evaluating invalid markets.

Similarly we could challenge as invalid if `answer is None` (i.e. `is_predictable_binary(market_question) == False`), although this isn't using any OFV stuff.

Reply:
>  did this include any invalid markets?

Unfortunatelly no.

>  Wondering if it's a bad idea to do this

I guess we can? It's a valid output from OFV and there will be bond for people to fix it 🤷 

> Similarly we could challenge as invalid if answer is None

However, I wouldn't do this. Because if `asnwer is None` it can mean arbitrary exception. in OFV run(missing credits, pydantic model problem, etc.).

> is_predictable_binary(market_question) == False

That would mean bug in Olas' market-creator, maybe we should log it as error in this agent? But I wouldn't resolve it as invalid, because non-binary markets are valid markets, just not supported by us.

Reply:
Good points!

> Should we challenge to Invalid if factuality is None?

Okay, let's got for it!

Reply:
Updated 🙏 YOLO style!